src/Adapter/MSTest.TestAdapter/Execution/TestClassInfo.cs

```diff
@@ -387,7 +395,9 @@ internal UnitTestResult GetResultOrRunClassInitialize(ITestContext testContext,
                 PlatformServiceProvider.Instance.AdapterTraceLogger.LogWarning(Resource.STAIsOnlySupportedOnWindowsWarning);
             }
 
-            return DoRun();
+            UnitTestResult result = DoRun();
+            _cachedClassInitializeResult = result;
```

Comment:
The value set from this code path (non-STA) is not read. We only read in STA path. Do we want to do a check at the beginning that's common for STA and non-STA like this?

```csharp
if (_cachedClassInitializeResult is not null)
{
    return _cachedClassInitializeResult;
}

Reply:
It's returned on 366, we could do `if (IsClassInitializeExecuted || _...)` or even simplify as the if you described and invert the assert,

Reply:
@Evangelink Line 366 is for STA specifically, while line 399 is for non-STA.

Reply:
Nope. 366 is generic, 381 STA and here 399 for non-STA